go/store/prolly/tree_chunker.go

```diff
@@ -0,0 +1,444 @@
+		if err != nil {
+			return err
+		}
+		if ok && sc.cur.atNodeEnd() {
+			if sc.cur.parent != nil {
+
+				if sc.cur.parent.compare(next.parent) < 0 {
+					// Case (4): We stopped consuming items on this level before entering
+					// the Node referenced by |next|
+					reachedNext = false
+				}
+
+				// Note: Logically, what is happening here is that we are consuming the
+				// item at the current level. Logically, we'd call sc.cur.forward(),
+				// but that would force loading of the next Node, which we don't
+				// need for any reason, so instead we forward the parent and take care
+				// not to allow it to step outside the Node.
+				_, err := sc.cur.parent.advanceInBounds(ctx)
+				if err != nil {
+					return err
+				}
+
+				// Invalidate this cursor, since it is now inconsistent with its parent
+				sc.cur.parent = nil
+				sc.cur.nd = nil
+			}
+
+			break
+		}
+
+		if _, err := sc.cur.advance(ctx); err != nil {
+			return err
+		}
+	}
+
+	if sc.parent != nil && next.parent != nil {
+		err := sc.parent.advanceTo(ctx, next.parent)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	sc.cur = next
+	if !reachedNext {
+		err := sc.resume(ctx) // Case (4)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (sc *treeChunker) Skip(ctx context.Context) error {
+	_, err := sc.cur.advance(ctx)
+	return err
+}
+
+// Append adds a new key-value pair to the chunker, validating the new pair to ensure
+// that chunks are well-formed. Key-value pairs are appended atomically a chunk boundary
+// may be made before or after the pair, but not between them.
+func (sc *treeChunker) Append(ctx context.Context, key, value nodeItem) (bool, error) {
+	// When adding new key-value pairs to an in-progress chunk, we must enforce 3 invariants
+	// (1) Key-value pairs are stored in the same Node.
+	// (2) The total size of a Node's data cannot exceed |maxNodeDataSize|.
+	// (3) Internal Nodes (level > 0) must contain at least 2 key-value pairs (4 node items).
+	//     Infinite recursion can occur if internal nodes contain a single metaPair with a key
+	//     large enough to trigger a chunk boundary. Forming a chunk boundary after a single
+	//     key will lead to an identical metaPair in the next level in the tree, triggering
+	//     the same state infinitely. This problem can only occur at levels 2 and above,
+	//     but we enforce this constraint for all internal nodes of the tree.
+
+	// constraint (3)
+	degenerate := !sc.isLeaf() && len(sc.current) < metaPairCount*2
+
+	// constraint (2)
+	overflow := false
+	sum := sc.currSz + uint64(len(key)+len(value))
+	if sum >= maxNodeDataSize {
+		overflow = true
+	}
+
+	if overflow && degenerate {
+		// Constraints (2) and (3) are in conflict
+		panic("impossible node")
+	}
+
+	if overflow {
+		// Enforce constraints (1) and (2):
+		//  |key| and |value| won't fit in this chunk,
+		//  force a boundary here and pass them as leftovers
+		err := sc.handleChunkBoundary(ctx, key, value)
```

Comment:
In the "overflow" case, we're adding `key` and `value` as "leftovers", and adding them again on line 235.
Needs a fix